src/core/thread/router_table.cpp

```diff
@@ -219,38 +219,26 @@ Router *RouterTable::Allocate(void)
         if (!IsAllocated(routerId) && mRouterIdReuseDelay[routerId] == 0)
         {
             numAvailable++;
-        }
-    }
-
-    VerifyOrExit(mActiveRouterCount < Mle::kMaxRouters && numAvailable > 0);
 
-    // choose available router id at random
-    freeBit = Random::NonCrypto::GetUint8InRange(0, numAvailable);
+            // Randomly select a router ID as we iterate through the
+            // list using Reservoir algorithm: We replace the
+            // selected ID with current entry in the list with
+            // probably `1/numAvailable`.
 
-    // allocate router
-#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
-    for (uint8_t routerId = mMinRouterId; routerId <= mMaxRouterId; routerId++)
-#else
-    for (uint8_t routerId = 0; routerId <= Mle::kMaxRouterId; routerId++)
-#endif
-    {
-        if (IsAllocated(routerId) || mRouterIdReuseDelay[routerId] > 0)
-        {
-            continue;
+            if (Random::NonCrypto::GetUint8InRange(0, numAvailable) == 0)
+            {
+                selectedRouterId = routerId;
+            }
         }
+    }
 
-        if (freeBit == 0)
-        {
-            rval = Allocate(routerId);
-            OT_ASSERT(rval != nullptr);
-            ExitNow();
-        }
+    VerifyOrExit((mActiveRouterCount < Mle::kMaxRouters) && (selectedRouterId != Mle::kInvalidRouterId));
```

Comment:
nit: should we perform the `mActiveRouterCount < Mle::kMaxRouters` check before the for-loop? It avoids having to generate a random value a number of times.